### ISSUE TITLE
T5Gemma training not working

### ISSUE THREAD

**Issue by NilsHellwig** [huggingface/transformers#39656]
### System Info

- `transformers` version: 4.53.3
- Platform: Linux-6.11.0-29-generic-x86_64-with-glibc2.39
- Python version: 3.11.13
- Huggingface_hub version: 0.33.5
- Safetensors version: 0.5.3
- Accelerate version: 1.9.0
- Accelerate config:    not found
- DeepSpeed version: not installed
- PyTorch version (accelerator?): 2.7.1+cu126 (CUDA)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: <fill in>
- Using GPU in script?: <fill in>
- GPU type: NVIDIA GeForce RTX 4090

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

1. run the code:

```
#!/usr/bin/env python3
"""
T5Gemma Fine-tuning Example for Classification Task
Uses IMDB sentiment classification dataset
"""

from datasets import load_dataset, concatenate_datasets
import numpy as np
import torch
from sklearn.metrics import accuracy_score, precision_recall_fscore_support
from transformers import (
    AutoTokenizer,
    AutoModelForSeq2SeqLM,
    Seq2SeqTrainer,
    Seq2SeqTrainingArguments,
    EvalPrediction
)

# Device detection
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Using device: {device}")

# Load model and tokenizer
model_name = "google/t5gemma-b-b-ul2"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSeq2SeqLM.from_pretrained(
    model_name, 
    attn_implementation="eager",
    device_map="auto",
    torch_dtype=torch.bfloat16
)

# Load IMDB dataset for sentiment classification
print("Loading IMDB dataset...")
dataset = load_dataset("imdb")

# Use smaller subset for faster training
# Create balanced subsets with equal positive and negative samples
train_pos = dataset["train"].filter(lambda x: x["label"] == 1).select(range(500))
train_neg = dataset["train"].filter(lambda x: x["label"] == 0).select(range(500))
train_subset = concatenate_datasets([train_pos, train_neg])

test_pos = dataset["test"].filter(lambda x: x["label"] == 1).select(range(50))
test_neg = dataset["test"].filter(lambda x: x["label"] == 0).select(range(50))
test_subset = concatenate_datasets([test_pos, test_neg])

def preprocess(example):
    """Preprocessing for sentiment classification"""
    # Format input text for T5-style classification
    input_text = f"classify sentiment: {example['text'][:512]}"  # Truncate long texts
    target_text = "positive" if example["label"] == 1 else "negative"
    target_text = target_text + tokenizer.eos_token
    
    # Tokenize
    model_inputs = tokenizer(input_text, max_length=512, truncation=True, padding="max_length")
    labels = tokenizer(target_text, max_length=10, truncation=True, padding="max_length")
    
    # Prepare labels for loss calculation
    labels_ids = [label if label != tokenizer.pad_token_id else -100 for label in labels["input_ids"]]
    model_inputs["labels"] = labels_ids
    
    return model_inputs

def compute_metrics(eval_pred: EvalPrediction):
    """Compute accuracy and F1 score"""
    predictions, labels = eval_pred
    
    # Decode predictions and labels
    pred_str = tokenizer.batch_decode(predictions, skip_special_tokens=True)
    labels = np.where(labels != -100, labels, tokenizer.pad_token_id)
    label_str = tokenizer.batch_decode(labels, skip_special_tokens=True)
    
    # Convert sentiment strings to binary labels
    pred_labels = [1 if p.strip().lower() == "positive" else 0 for p in pred_str]
    true_labels = [1 if l.strip().lower() == "positive" else 0 for l in label_str]
    print(pred_labels)
    
    # Calculate metrics
    accuracy = accuracy_score(true_labels, pred_labels)
    precision, recall, f1, _ = precision_recall_fscore_support(true_labels, pred_labels, average='binary')
    
    return {
        "accuracy": accuracy,
        "f1": f1,
        "precision": precision,
        "recall": recall
    }

# Tokenize datasets
print("Tokenizing datasets...")
train_tokenized = train_subset.map(preprocess, remove_columns=train_subset.column_names)
test_tokenized = test_subset.map(preprocess, remove_columns=test_subset.column_names)

# Training arguments
training_args = Seq2SeqTrainingArguments(
    output_dir="./t5gemma-imdb-finetuned",
    eval_strategy="steps",
    eval_steps=500,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    num_train_epochs=1,
    save_strategy="steps",
    save_steps=500,
    save_total_limit=2,
    load_best_model_at_end=True,
    metric_for_best_model="accuracy",
    predict_with_generate=True,
    bf16=True,
    remove_unused_columns=False,  # Important for T5Gemma!
    logging_steps=100,
    warmup_steps=100,
    learning_rate=5e-5,
)

# Initialize trainer
trainer = Seq2SeqTrainer(
    model=model,
    args=training_args,
    train_dataset=train_tokenized,
    eval_dataset=test_tokenized,
    compute_metrics=compute_metrics,
)

if __name__ == "__main__":
    print("Starting training...")
    trainer.train()
    
    print("Evaluating on test set...")
    test_results = trainer.evaluate(test_tokenized)
    print(f"Test Accuracy: {test_results['eval_accuracy']:.4f}")
    print(f"Test F1: {test_results['eval_f1']:.4f}")
    
    # Save model
    trainer.save_model("./t5gemma-imdb-final")
    print("Model saved!")

    # Example inference
    print("\nExample inference:")
    input_text = "classify sentiment: This movie was absolutely fantastic! Great acting and storyline."
    inputs = tokenizer(input_text, return_tensors="pt").to(model.device)
    outputs = model.generate(**inputs, max_new_tokens=10)
    prediction = tokenizer.decode(outputs[0], skip_special_tokens=True)
    print(f"Input: {input_text}")
    print(f"Prediction: {prediction}")

```

### Expected behavior

T5Gemma model should learn. Prediction is empty regardless of hyperparameters. T5-base works fine.

**Comment by tanuj-rai** [comment]
Hi @NilsHellwig, can you please tell me what error you are getting?

**Comment by NilsHellwig** [comment]
> Hi @NilsHellwig, can you please tell me what error you are getting?

Thank you for your response. The model is not training, always predicting the same class. Works with t5-base but not with t5gemma. I would recommend for you to test it yourself. It's my entire code.

**Comment by NilsHellwig** [comment]
@tanuj-rai

**Comment by jadermcs** [comment]
@NilsHellwig I had the same issue in #39521
The problem is that the Tokenizer don't prepare the inputs for T5Gemma model, so you need to do it manually.
Check the official usage for some insights:
https://github.com/google-gemini/gemma-cookbook/blob/main/Research/%5BT5Gemma%5DExample.ipynb

**Comment by NilsHellwig** [comment]
@jadermcs So isn't it possible to use it without the kaggle toolkit. That would be really sad.

**Comment by Rocketknight1** [comment]
@jadermcs can you clarify which bits are missing from the Tokenizer? Can we update the T5Gemma tokenizers to resolve this?

**Comment by jadermcs** [comment]
@NilsHellwig @Rocketknight1 
There's a bug in the decoder that is not adding the EOS token.

The forward of the model is shifting the tokens to the right in the decoder even when you explicitly define the input and labels for the decoder.

If I do:
```python
def preprocess(example):
    input_text = (
        "word: {word} identical in sentence 1: {sentence1} sentence 2: {sentence2}"
    ).format(
        word=example["lemma"],
        sentence1=example["sentence1"],
        sentence2=example["sentence2"],
    )
    target_text = "true" if example["labels"] == 1 else "false"
    target_text += tokenizer.eos_token

    # Tokenize inputs and targets
    model_inputs = tokenizer(
        input_text, max_length=128, truncation=True, padding="max_length"
    )
    labels = tokenizer(target_text, max_length=5, truncation=True, padding="max_length")
    model_inputs["decoder_input_ids"] = [tokenizer.bos_token_id] + labels["input_ids"][:-1]
    model_inputs["decoder_attention_mask"] = [1] + labels["attention_mask"][:-1]

    model_inputs["labels"] = [
        label if label != tokenizer.pad_token_id else -100 for label in labels["input_ids"]
    ]
    return model_inputs

```
I have:

```python
prediction = ['<bos>false<eos><pad>', '<bos>false<eos><pad>', '<bos>false<eos><pad>']
labels = ['false<eos><pad><pad>', 'false<eos><pad><pad>', 'true<eos><pad><pad>']
```

**Comment by jadermcs** [comment]
> [@jadermcs](https://github.com/jadermcs) can you clarify which bits are missing from the Tokenizer? Can we update the T5Gemma tokenizers to resolve this?

I thought I knew the source of the bug, but I couldn't confirm my guess. I still suspect is the interplay between tokens shift and how the loss is calculated.
